src/cache/cache_las.c

```diff
@@ -0,0 +1,390 @@
+			WT_WITHOUT_DHANDLE(session, ret =
+			    __las_cursor_create(session, &session->las_cursor));
+			WT_RET(ret);
+		}
+
+		*cursorp = session->las_cursor;
+	} else {
+		/* Lock the shared lookaside cursor. */
+		__wt_spin_lock(session, &conn->las_lock);
+
+		*cursorp = conn->las_cursor;
+	}
+
+	/* Turn eviction off. */
+	if (*reset_evict)
+		F_SET(session, WT_SESSION_NO_EVICTION);
+
+	return (0);
+}
+
+/*
+ * __wt_las_cursor_close --
+ *	Discard a lookaside cursor.
+ */
+int
+__wt_las_cursor_close(
+	WT_SESSION_IMPL *session, WT_CURSOR **cursorp, int reset_evict)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_CURSOR *cursor;
+	WT_DECL_RET;
+
+	conn = S2C(session);
+
+	if ((cursor = *cursorp) == NULL)
+		return (0);
+	*cursorp = NULL;
+
+	/* Reset the cursor. */
+	ret = cursor->reset(cursor);
+
+	/*
+	 * We turned off eviction while the lookaside cursor was in use, restore
+	 * the session's flags.
+	 */
+	if (reset_evict)
+		F_CLR(session, WT_SESSION_NO_EVICTION);
+
+	/*
+	 * Eviction and sweep threads have their own lookaside table cursors;
+	 * else, unlock the shared lookaside cursor.
+	 */
+	if (!F_ISSET(session, WT_SESSION_LOOKASIDE_CURSOR))
+		__wt_spin_unlock(session, &conn->las_lock);
+
+	return (ret);
+}
+
+/*
+ * __wt_las_sweep --
+ *	Sweep the lookaside table.
+ */
+int
+__wt_las_sweep(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_CURSOR *cursor;
+	WT_DECL_ITEM(las_addr);
+	WT_DECL_ITEM(las_key);
+	WT_DECL_RET;
+	WT_ITEM *key;
+	uint64_t cnt, las_counter, las_txnid;
+	uint32_t las_id;
+	int notused, reset_evict;
+
+	conn = S2C(session);
+	cursor = NULL;
+	key = &conn->las_sweep_key;
+	reset_evict = 0;		/* [-Werror=maybe-uninitialized] */
+
+	WT_ERR(__wt_scr_alloc(session, 0, &las_addr));
+	WT_ERR(__wt_scr_alloc(session, 0, &las_key));
+
+	WT_ERR(__wt_las_cursor(session, &cursor, &reset_evict));
+
+	/*
+	 * If we're not starting a new sweep, position the cursor using the key
+	 * from the last call (we don't care if we're before or after the key,
+	 * just roughly in the same spot is fine).
+	 */
+	if (conn->las_sweep_call != 0 && key->data != NULL) {
+		__wt_cursor_set_raw_key(cursor, key);
+		if ((ret =
+		    cursor->search_near(cursor, &notused)) == WT_NOTFOUND) {
+			WT_ERR(cursor->reset(cursor));
+			return (0);
+		}
+		WT_ERR(ret);
+	}
+
+	/*
+	 * The sweep server wakes up every 10 seconds (by default), it's a slow
+	 * moving thread. Try to review the entire lookaside table once every 5
+	 * minutes, or every 30 calls.
+	 *
+	 * The reason is because the lookaside table exists because we're seeing
+	 * cache/eviction pressure (it allows us to trade performance and disk
+	 * space for cache space), and it's likely lookaside blocks are being
+	 * evicted, and reading them back in doesn't help things. A trickier,
+	 * but possibly better, alternative might be to review all lookaside
+	 * blocks in the cache in order to get rid of them, and slowly review
+	 * lookaside blocks that have already been evicted.
```

Comment:
Good point -- do we ever want to cache lookaside table pages we read? I'm guessing not, there's no reason to believe they'll be useful (unless we have a workload where reading one page that needs LAS records implies we're likely to read other pages that need LAS records, and that seems like a stretch).